text/0000-generic-integers.md

```diff
@@ -0,0 +1,437 @@
+
+## Bound-based generalisation
+
+Generalising integers over their number of bits is certainly a very logical way to generalise integers *for computers*, but generalising based upon bounds is a very natural way for humans to do it, *and* more general. For example, instead of `u<N>` and `i<N>` types, we could get away with just one type,
+`int<A..=B>`. This would be more powerful than the original: for example, a percentage could be represented exactly as `int<0..=100>`. Whether an integer is signed simply depends on whether its lower bound is negative.
+
+The primary reason for leaving this out is… well, it's a lot different from the existing integer types in the language. Additionally, such a proposal can coexist with this one, since there's nothing stopping us from making an `int<A..=B>` type in the future such that `u<N>` and `i<N>` are equivalent to the correct range of integers.
+
+## Integer traits
+
+Previously, Rust had traits which generalized the integer types and their methods, but these were ultimately removed in favour of inherent methods. Going with a generic `u<N>` over an `Int` trait would avoid the problem of determining which methods are suited for these traits; instead, the `u<N>` type would have all of them.
+
+Additionally, having these traits does not allow non-power-of-two `u<N>` outright, and so, this feature is a strict superset of that.
+
+Having separate `u<N>` and `i<N>` types also solves the problem of generalising signed and unsigned integers: everything would require one impl for signed integers, and one for unsigned. This would ensure that these types have exactly the same behaviour for all `N`, only differing in the upper bound on the number of bits.
+
+## Offering as a library
+
+This was the main proposal last time this RFC rolled around, and as we've seen, it hasn't really worked.
+
+Crates like [`u`], [`bounded-integer`], and [`intx`] exist, but they come with their own host of problems:
+
+* None of these libraries can easily unify with the existing `uN` and `iN` types.
+* Generally, they require a lot of unsafe code to work.
+* These representations tend to be slower and less-optimized than compiler-generated versions.
+* They still require you to generalise integer types with macros instead of const generics.
+
+A library solution really doesn't feel like the right option here. While libraries can create general integer *traits* to work over all of the existing `uN` and `iN` types, they can't easily make generic integer types.
+
+[`u`]: https://docs.rs/u
+[`bounded-integer`]: https://docs.rs/bounded-integer
+[`intx`]: https://docs.rs/intx
+
+## Going without
+
+This is always an option, but hopefully it seems like a worse option after all that's been said so far.
+
+# Prior art
+[prior-art]: #prior-art
+
+* [The previous RFC.][#2581]
+* [Zulip RFC revival topic.][Zulip]
+* [Generic integers in Zig.][Zig]
+* [Generic integers in C23.][`_BitInt`]
+* Probably several others discussions I'm missing.
```

Comment:
is the recent [swift](https://forums.swift.org/t/integer-generic-parameters/74181) proposal relevant here?

Reply:
Probably! Thank you for letting me know; I'll add it to the list.

Reply:
Actually, no, it's not; I totally should have read that before updating the RFC. It appears to just be an RFC for const generics, which Rust already has, and not generic integers.

Reply:
apologies, I agree that it actually is only const generics, rather than these generic integers.... oops

Reply:
All good! Any blame I could assign to you for not reading I could also assign to myself, so…